File: structural/Java/Proxy.java
abstract class Image {

	protected String name;

	public abstract void display();

}

class RealImage extends Image {

	public RealImage(String imageName) {
		this.name = imageName;
		this.loadFromDisk(this.name);
	}

	@Override
	public void display() {
		System.out.println("Displaying " + this.name + " from memory.\n");
	}

	private void loadFromDisk(String imageName) {
		System.out.println("Laoding image: " + imageName + " from disk.");
	}

}

class ProxyImage extends Image {

	private RealImage realImage;

	public ProxyImage(String imageName) {
		this.name = imageName;
	}

	@Override
	public void display() {
		if (this.realImage == null) {
			this.realImage = new RealImage(this.name);
		}
		this.realImage.display();
	}

}

public class Main {

	public static void main(String[] args) {
		System.out.println(":: Proxy ::\n");
		
		Image image = new ProxyImage("test_10mb.jpg");
		
		// image will be loaded from disk.
		image.display();
		
		// image will not be loaded from disk.
		image.display();
	}

}

File: structural/javascript/Proxy.js
// abstract class
class Image {

  constructor() {
    if (this.constructor === Image) {
      throw new TypeError('Cannot construct Abstract instances directly');
    }

    if (this.display === Image.prototype.display) {
      throw new TypeError('Must override method "display()"');
    }

    this.name = '';
  }

  display() {
    throw new TypeError('Do not call abstract method "display()" from child.');
  }

}

class RealImage extends Image {

  constructor(imageName) {
    super();
    this.name = imageName;
    this.loadFromDisk(this.name);
  }

  // Implements the abstract parent method
  display() {
    console.log('Displaying ' + this.name + ' from memory.\n');
  }

  loadFromDisk(imageName) {
    console.log('Loading image: ' + imageName + ' from disk.');
  }

}

class ProxyImage extends Image {

  constructor(imageName) {
    super();
    this.realImage = null;
    this.name = imageName;
  }

  // override
  display() {
    if (!this.realImage) {
      this.realImage = new RealImage(this.name);
    }
    this.realImage.display();
  }

}

console.log(':: Proxy ::\n');

const image = new ProxyImage('test_10mb.jpg');

// image will be loaded from disk.
image.display();

// image will not be loaded from disk.
image.display();